### ISSUE TITLE
Adding observable to DTO class properties modifies property name and has stringify issues

### ISSUE THREAD

**Issue by rossvegas** [aurelia/framework#818]
**I'm submitting a bug report**

* **Library Version:**
1.1.4

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Node Version:**
8.2.1

* **NPM Version:**
5.3.0

* **Webpack**
webpack 2.5.1

* **Browser:**
Chrome 60.0.3112.101

* **Language:**
TypeScript 2.4.2

**Current behavior:**

Hi I am creating a little task tracking application to teach myself Aurelia and have come across what I think is a little bug(maybe).

I have the following class as part of a experimental very basic task tracker application

```
export class Goal {
    public static fromObject(src) {
        return Object.assign(new Goal(), src);
    }    
    public Id: number = 0;
    public CategoryId: number | null = null;
    public Comment: string | null = null;
    public Done: boolean = false;
    public DueDate: Date | null = null;
    public ModifiedById: number = 0;
    public Note: string = "";
    public Urgent: boolean = false;
    public UserId: number = 0;
}
```
this is bound to a UI which updates and saved via the fetch client perfectly.

The object prior to serialization looks like this in the chrome debugger:
```
Goal {Id: 2, …}
Category:null
CategoryId:2
Comment:"For Monday"
Dirty:false
Done:false
DueDate:"2017-08-17"
Id:2
ModifiedBy:null
ModifiedById:3
Note:"Wash Clothes"
Urgent:true
User:null
UserId:1
__observers__
:
{Done: e, Note: e, Comment: e, DueDate: e, CategoryId: e}
get CategoryId:ƒ ()
set CategoryId:ƒ ()
get Comment:ƒ ()
set Comment:ƒ ()
get Done:ƒ ()
set Done:ƒ ()
get DueDate:ƒ ()
set DueDate:ƒ ()
get Note:ƒ ()
set Note:ƒ ()
__proto__
:
Object
```
the json generated from a JSON.Stringify looks like this
```
"{"Id":2,"CategoryId":2,"Comment":"For Monday","Done":false,"DueDate":"2017-08-17","ModifiedById":3,"Note":"Wash Clothes","Urgent":true,"UserId":1,"User":null,"Category":null,"ModifiedBy":null,"Dirty":false}"
```
At this point everything is awesome!

I then decided to add observable to the class against a few properties e.g.

```
import { bindable, observable } from "aurelia-framework";

export class Goal {

    public static fromObject(src) {
        return Object.assign(new Goal(), src);
    }

    public Dirty: boolean;
    public Id: number = 0;
    public CategoryId: number | null = null;
    @observable public Comment: string | null = null;
    public Done: boolean = false;
    @observable public DueDate: Date | null = null;
    public ModifiedById: number = 0;
    @observable public Note: string = "";
    public Urgent: boolean = false;
    public UserId: number = 0;

    public CommentChanged() {
        this.Dirty = true;
    }

    public DueDateChanged() {
        this.Dirty = true;
    }

    public NoteChanged() {
        this.Dirty = true;
    }
```
the object in chrome now looks like this
```
Goal {Id: 2, Dirty: false, ModifiedById: 3, …}
Category:null
CategoryId:2
Comment:"For Monday"
Dirty:false
Done:false
DueDate:"2017-08-17"
Id:2
ModifiedBy:null
ModifiedById:3
Note:"Wash Clothes"
Urgent:true
User:null
UserId:1
_Comment:"For Monday"
_DueDate:"2017-08-17"
_Note:"Wash Clothes"
__observers__
:
{Done: e, Note: e, _Note: e, Comment: e, _Comment: e, …}
get CategoryId:ƒ ()
set CategoryId:ƒ ()
get Done:ƒ ()
set Done:ƒ ()
get _Comment:ƒ ()
set _Comment:ƒ ()
get _DueDate:ƒ ()
set _DueDate:ƒ ()
get _Note:ƒ ()
set _Note:ƒ ()
__proto__
:
Object
```
Notice in this object the property names for Note, Comment and Due Date all have an underscore now where previously they didn't.
Stringifying this object has this result.
```
"{"Id":2,"CategoryId":2,"Dirty":false,"Done":false,"ModifiedById":3,"Urgent":true,"UserId":1,"User":null,"Category":null,"ModifiedBy":null}"
```
As you can see now the Note, Comment and Due Date fields have been excluded from the string resulting in a loss of data.

* **Expected/desired behavior:**

I would expect that adding observable would not result in the property being renamed and excluded from the stringified object.

* **What is the expected behavior?**

Getters and setters retain their names and are serialized just as they are when they are not marked with observable

**Comment by EisenbergEffect** [comment]
@jdanyow Can you look into this? Seems like there may be an issue with the enumerable aspect of the generated properties.

**Comment by bigopon** [comment]
@EisenbergEffect @jdanyow I think this is an easy fix

If it's decorated on class, then the above is expected behavior
Else just need to change descriptor enumerable to true

**Comment by bigopon** [comment]
@rossvegas There is a work around for you in the mean time, add `toJSON` method to your model class

```js
  toJSON() {
    return {
        data1: 1,
        // ...
    }
  }
```

`JSON.stringify` will look for `toJSON` method on the object, that's why you get ISO format when calling `JSON.stringify` on a `Date` instance, if you haven't noticed.

@EisenbergEffect Since there is a work around for this, maybe no need to change this as it could be considered breaking change ?

**Comment by rossvegas** [comment]
Thanks @bigopon  for the workaround that works fine.

As far as not changing it because it could be a breaking change I think it is worth the change, the resultant effect will be that properties decorated with observable will still be there instead of excluded when calling stringify, this being just an addition of data to the string not a modification or deletion I don't think the impact will be too great to existing code bases. Having said that I am just starting with Aurelia and I don't now what else might be affected.

Just my two cents but I don't think that decorating a property should change behavior in serialization and I think the current behavior its is terribly confusing for anyone trying to adopt the platform.

**Comment by bigopon** [comment]
@rossvegas  Oh, actually what I said was incorrect, as decorating a field will make it go to the prototype. so it will never go to json stringify. Seems like you have to use `toJSON`.

**Comment by bigopon** [comment]
@rossvegas Update: your issue can be solved by doing:

```ts
class MyClass {
 @observable() prop: string
}
```

instead of

```ts
class MyClass {
  @observable prop: string
}
```

the reason is typescript doesn't provide property descriptor when decorating without parens.

@EisenbergEffect I'm recreating type observable decorator, and it will mitigate this issue

**Comment by Alexander-Taran** [comment]
Maybe can be closed

**Comment by EisenbergEffect** [comment]
Should be fixed now since the props are no longer enumerable.

**Comment by mipod** [comment]
I know that this topic is closed but I don't know where else to ask. I  have a problem with the @observable decorator from Aurelia. When I try to call JSON.stringify on a model class that has an property with the decorator it will be eliminated from the JSON data. For example:

`export class ProductModel {
  productId: string;
  @observable quantity: number;
}`

I initialize a variable: 
`var product : ProductModel = new ProductModel();
product.productId = "test";
product.quantity = 1;
`

and then call JSON.stringify on it it eliminates the quantity property. 
`JSON.stringify(product);

// output: { productId : "test" }`

Am I using it wrong or is this a bug? 

Thanks for the help

**Comment by rossvegas** [comment]
I think I ended up putting parenthesis after the observable, so 
export class ProductModel { productId: string; @observable quantity: number; }
became
export class ProductModel { productId: string; @observable() quantity: number; }

Give that a go and let us know if that works

**Comment by mipod** [comment]
@rossvegas I have tried  putting "()" after the observable keyword with no luck. When it hits the JSON,stringify it just ignores that field, like it never existed...

**Comment by rossvegas** [comment]
Ok so I remember what I did, its the best I could come up with but it's a bit ordinary.

using @observable you need to overwrite toJson. I wanted to track dirty in the class so it would be 

```
export class ProductModel { 

	@observable productId: string; 
	@observable quantity: number;
	public dirty: boolean;

	public productIdChanged() {
        this.setDirty();
    }

    public quantityChanged() {
        this.setDirty();
    }

	public setDirty() {
        this.dirty = true;
    }

	public toJSON() {
	return {
		productId : this.productId,
		quantity : this.quantity
	}
}
```

depending on what you are trying to do you could drop observable and try computed from for a particular property. That way you wont have to decorate the properties with @observable and they will strigify just fine.

http://aurelia.io/docs/binding/computed-properties#introduction

**Comment by mipod** [comment]
@rossvegas  thank for the response. I ended up doing something similar to what you suggested. I modified the class to look like this:
`export class ProductForReservationModel {
  productId: string;
  @observable() observableQuantity: number;
  quantity: number;
  tickets: TicketForReservationModel[];

  observableQuantityChanged(newValue) {
    this.quantity = newValue;
  }

}`

I use the observable because I use it to trigger different behavior when that field has changed value. 
Thank you again for your answer.
